pageserver/src/tenant.rs

```diff
@@ -422,13 +422,46 @@ impl Tenant {
             init_order,
             CreateTimelineCause::Load,
         )?;
-        let new_disk_consistent_lsn = timeline.get_disk_consistent_lsn();
+        let disk_consistent_lsn = timeline.get_disk_consistent_lsn();
+        assert_eq!(
+            disk_consistent_lsn,
+            up_to_date_metadata.disk_consistent_lsn(),
+            "these are used interchangeably; need to move layer map loading before Timeline creation to avoid drifting"
+        );
         anyhow::ensure!(
-            new_disk_consistent_lsn.is_valid(),
+            disk_consistent_lsn.is_valid(),
             "Timeline {tenant_id}/{timeline_id} has invalid disk_consistent_lsn"
         );
+
+        let index_part = remote_startup_data.as_ref().map(|x| &x.index_part);
+
+        // FIXME: it seems that we shouldn't have "mutable init" for RemoteTimelineClient either
+        if let Some(index_part) = index_part {
+            timeline
+                .remote_client
+                .as_ref()
+                .unwrap()
+                .init_upload_queue(index_part)?;
+        } else if self.remote_storage.is_some() {
+            // No data on the remote storage, no local layers, local metadata file.
+            //
+            // TODO https://github.com/neondatabase/neon/issues/3865
+            // Currently, console does not wait for the timeline data upload to the remote storage
+            // and considers the timeline created, expecting other pageserver nodes to work with it.
+            // Branch metadata upload could get interrupted (e.g pageserver got killed),
+            // hence any locally existing branch metadata with no remote counterpart should be uploaded,
+            // otherwise any other pageserver won't see the branch on `attach`.
+            //
+            // After the issue gets implemented, pageserver should rather remove the branch,
+            // since absence on S3 means we did not acknowledge the branch creation and console will have to retry,
+            // no need to keep the old files.
```

Comment:
I've actually fixed the issue here but haven't done the remove the branch... I never discovered this comment.

Reply:
I'll edit the text to have link to https://github.com/neondatabase/neon/issues/5075

Reply:
0ee01b37f

Reply:
I didn't see your comment:
> Yeah, stumbled across this as well. I think it's ok to remove this comment.

We clearly do not have the cleanup part implemented. Do you disagree?

Reply:
Cleanup part := 

```
/ After the issue gets implemented, pageserver should rather remove the branch,
            // since absence on S3 means we did not acknowledge the branch creation and console will have to retry,
            // no need to keep the old files.
```

I disagree with that. Console is required to retry create requests until they succeed, so, if we have it locally but not remotely, why not re-do the upload.

Reply:
Ok, makes sense because console cannot stop retrying operations, so we should always get to 409 Conflict response from create_timeline which has waited for uploads to finish. So we need to make sure even the empty index file is being uploaded.

b08c60cef adjusts the comment, removes the link to 5075.